src/metrics/test/metrics_test.cpp

```diff
@@ -502,5 +508,122 @@ TEST_CASE("Registry counts the number of workers") {
     REQUIRE(metrics.getWorkerCount("actor2", "op1") == 2);
 }
 
+TEST_CASE("Events stream to gRPC") {
+    using EventVec = std::vector<poplar::EventMetrics>;
+
+    auto compareEventsAndClear = [](const EventVec& events_in) {
+        internals::v2::MockStreamInterface interface;
+        REQUIRE(events_in.size() == interface.events.size());
+        for (int i = 0; i < events_in.size(); i++) {
+            REQUIRE(google::protobuf::util::MessageDifferencer::Equals(events_in[i],
+                                                                       interface.events[i]));
+            // REQUIRE(2 == 1);
+        };
+        interface.events.clear();
+    };
+
+    SECTION("Empty stream") {
+        RegistryClockSourceStub::reset();
+        auto stream =
+            internals::v2::EventStream<RegistryClockSourceStub, internals::v2::MockStreamInterface>{
+                1, "TestName", 1, "/test/prefix"};
+
+        EventVec expected;
+        compareEventsAndClear(expected);
+    }
+
+    SECTION("Two events") {
+        RegistryClockSourceStub::reset();
+        auto stream =
+            internals::v2::EventStream<RegistryClockSourceStub, internals::v2::MockStreamInterface>{
+                1, "EventName", 9, "/test/prefix"};
+        EventVec expected;
+
+
+        // First Event
+        RegistryClockSourceStub::advance(std::chrono::microseconds(5));
+        {
+            OperationEventT<RegistryClockSourceStub> event(
+                2,
+                3,
+                7,
+                0,
+                Period<RegistryClockSourceStub>{std::chrono::microseconds(5)},
+                OutcomeType::kSuccess);
+            stream.addAt(RegistryClockSourceStub::now(), event, 1);
+
+            // Streamed Event
+            // ---------------
+            // Expected event
+
+            poplar::EventMetrics metric;
+            metric.set_name("EventName");
+            metric.set_id(1);
+
+            metric.mutable_time()->set_seconds(0);
+            metric.mutable_time()->set_nanos(5000);
+
+            metric.mutable_timers()->mutable_duration()->set_seconds(0);
+            metric.mutable_timers()->mutable_duration()->set_nanos(5000);
+            metric.mutable_timers()->mutable_total()->set_seconds(0);
+            metric.mutable_timers()->mutable_total()->set_nanos(5000);
+
+            metric.mutable_counters()->set_number(2);
+            metric.mutable_counters()->set_ops(3);
+            metric.mutable_counters()->set_size(7);
+            metric.mutable_counters()->set_errors(0);
+
+            metric.mutable_gauges()->set_failed(false);
+            metric.mutable_gauges()->set_workers(1);
+            metric.mutable_gauges()->set_state(9);
+
+            expected.push_back(metric);
+        }
```

Comment:
Feels like we've got good happy-path testing, but personally I would be a bit afraid to make changes to the code with this testing.

Any way to assert:

1. the filenames being generated
2. what happens when two actors (threads with same Actor,Operation) send and/or get collectors?
3. behavior when the gRPC side of things barfs
4. behavior when metrics formats are other values
5. ensure that when an actor fails to call .success or .fail that we do/don't report to grpc?


It could well be that this testing is hard given the current design or testing libraries. If that's the case we can certainly punt on adding more un-happy-path tests and do them whenever we rip out csv metrics and/or when we add a better story for mocking. If that's the case, let's write a few sentences listing the cases that would be interesting to test but that are hard to test and make sure we point to those words in the places in the code that are currently untested. This will help future maintainers set their expectations about what the tests are or aren't doing.

Reply:
plz add a note about cases that aren't covered (maybe including some of my examples here). We'll want to circle back to add them once we remove CSV

Reply:
Added a comment above the `TEST_CASE` explaining that, hopefully exhaustive enough.